ports/raspberrypi/bindings/tca9555r/__init__.c

```diff
@@ -0,0 +1,743 @@
+    uint8_t reg_and_data[2] = { POLARITY_PORT0, polarity_state };
+    common_hal_busio_i2c_write(i2c, tca9555r_addresses[tca_index], reg_and_data, 2);
+    #if TCA9555R_LOCAL_MEMORY
+    tca9555r_polarity_state[LOW_BYTE(tca_index)] = polarity_state;
+    #endif
+}
+
+void tca_set_polarity_port_high(uint tca_index, uint8_t polarity_state) {
+    invalid_params_if(TCA9555R, tca_index >= TCA9555R_CHIP_COUNT);
+    busio_i2c_obj_t *i2c = common_hal_board_create_i2c(0);
+    uint8_t reg_and_data[2] = { POLARITY_PORT1, polarity_state };
+    common_hal_busio_i2c_write(i2c, tca9555r_addresses[tca_index], reg_and_data, 2);
+    #if TCA9555R_LOCAL_MEMORY
+    tca9555r_polarity_state[HIGH_BYTE(tca_index)] = polarity_state;
+    #endif
+}
+
+void tca_change_output_mask(uint8_t chip, uint16_t mask, uint16_t state) {
+    uint8_t low_mask = (uint8_t)(mask & 0xFF);
+    uint8_t low_state = (uint8_t)(state & 0xFF);
+    uint8_t high_mask = (uint8_t)(mask >> 8);
+    uint8_t high_state = (uint8_t)(state >> 8);
+    bool low_changed = low_mask > 0;
+    bool high_changed = high_mask > 0;
+    if (low_changed && high_changed) {
+        #if TCA9555R_LOCAL_MEMORY
+        uint16_t output_state = (tca9555r_output_state[HIGH_BYTE(chip)] << 8) | tca9555r_output_state[LOW_BYTE(chip)];
+        #else
+        uint16_t output_state = tca_get_output_port(chip);
+        #endif
+        uint16_t new_output_state = output_state;
+        new_output_state &= ~mask; // Clear the mask bits
+        new_output_state |= state; // Set the state bits
+        if (new_output_state != output_state) {
+            tca_set_output_port(chip, new_output_state);
+        }
+    } else if (low_changed) {
+        #if TCA9555R_LOCAL_MEMORY
+        uint8_t output_state = tca9555r_output_state[LOW_BYTE(chip)];
+        #else
+        uint8_t output_state = tca_get_output_port_low(chip);
+        #endif
+        uint8_t new_output_state = (output_state & ~low_mask) | low_state;
+        if (new_output_state != output_state) {
+            tca_set_output_port_low(chip, new_output_state);
+        }
+    } else if (high_changed) {
+        #if TCA9555R_LOCAL_MEMORY
+        uint8_t output_state = tca9555r_output_state[HIGH_BYTE(chip)];
+        #else
+        uint8_t output_state = tca_get_output_port_high(chip);
+        #endif
+        uint8_t new_output_state = (output_state & ~high_mask) | high_state;
+        if (new_output_state != output_state) {
+            tca_set_output_port_high(chip, new_output_state);
+        }
+    }
+}
+
+STATIC mp_obj_t tca_pin_get_number(mp_obj_t pin_obj) {
+    if (!mp_obj_is_type(pin_obj, &tca_pin_type)) {
+        mp_raise_TypeError_varg(translate("%q must be of type %q, not %q"), MP_QSTR_pin, tca_pin_type.name, mp_obj_get_type(pin_obj)->name);
+    }
+
+    mcu_pin_obj_t *pin = MP_OBJ_TO_PTR(pin_obj);
+    uint8_t tca_gpio = pin->number;
+    invalid_params_if(TCA9555R, tca_gpio >= TCA9555R_VIRTUAL_GPIO_COUNT);
```

Comment:
Are these `invalid_params_if` checking for internal errors or for errors in Python code? If the latter, they should also use an approprate function from argcheck.

Reply:
It's for internal errors I believe, checking that the pin number on the pin object is valid for the number of IO expanders being used (which should be unlikely to occur given the number of pins is hardcoded). None of the functions in argcheck look to be suitable.

Reply:
I've changed this over to an argcheck now. I hadn't realised the exact function that was referring to when you posted.